### ISSUE TITLE
Accessibility fix: Allow iOS browsers to auto-scale fonts based on user preference

### ISSUE THREAD

**Issue by chrislachance** [twbs/bootstrap#34107]
Please add font:-apple-system-body to the :root element CSS to allow iOS devices to auto-scale font sizes based on user preference set in Accessibility > Display & Text size.

**Comment by chrislachance** [comment]
Looks like it would need to exclude desktop Safari also.

@supports (-webkit-overflow-scrolling: touch) { /* Scope to iOS */
:root{font: -apple-system-body}}
}

**Comment by chrislachance** [comment]
See this Codepen for the full code example. https://codepen.io/chrislachance/pen/OJpmgrM

**Comment by ffoodd** [comment]
Wrapping a few resources for the record:
- [Adrian Roselli's "CSS and system fonts" post](https://adrianroselli.com/2015/07/css-and-system-fonts.html)
- [Greg Kraus wrote "Text Resizing for Web Pages in iOS using Dynamic Type"](https://www.interactiveaccessibility.com/blog/text-resizing-web-pages-ios-using-dynamic-type)
- [Safari 7 (!) changelog](https://developer.apple.com/library/archive/releasenotes/General/WhatsNewInSafari/Articles/Safari_7_0.html#//apple_ref/doc/uid/TP40014305-CH5-SW10)
- [CSSWG related discussion](https://github.com/w3c/csswg-drafts/issues/3708)

So to sum up my findings, I'd be in favor of trying this and heavily test the dedicated branch on iOS and MacOS browsers, since it's known to be tricky. 

BTW @chrislachance what kind of "weirdness" does desktop Safari, that worth mentioning in your pen?

**Comment by chrislachance** [comment]
Yeah, thanks. Jonathan Neal saw that without excluding Desktop Safari, the special Apple font treatment sets default font-size to 13px, instead of the standard 16px.

**Comment by ffoodd** [comment]
Do you have any link or resource to share regarding this? Currently unable to find any relate resource from Jonathan Neal on the topic :/

**Comment by chrislachance** [comment]
https://github.com/csstools/sanitize.css/issues/219

**Comment by ffoodd** [comment]
X-ref from other repos to keep an eye on the topic
- [Normalize.css #858](https://github.com/necolas/normalize.css/issues/858)
- [Modern Normalize #54](https://github.com/sindresorhus/modern-normalize/issues/54)
- [sanitize.css #219](https://github.com/csstools/sanitize.css/issues/219)

Not much discussions for now apart from Jonathan Neal on sanitize.css.

**Comment by chrislachance** [comment]
Looks like Apple is unwilling to change User Agent Stylesheet at the moment as well.

https://mobile.twitter.com/jonathandavis/status/1397996092214505472


On May 27, 2021 at 4:40 PM, Gaël Poupard ***@***.***) wrote:
X-ref from other repos to keep an eye on the topic
* Normalize.css #858 ( https://github.com/necolas/normalize.css/issues/858 )
* Modern Normalize #54 ( https://github.com/sindresorhus/modern-normalize/issues/54 )
* sanitize.css #219 ( https://github.com/csstools/sanitize.css/issues/219 )
Not much discussions for now apart from Jonathan Neal on sanitize.css.

**Comment by montygoldy** [comment]
"font:-apple-system-body" this solution not working for android chrome

**Comment by chrislachance** [comment]
It's Apple exclusive. Each Android already allows users to customize font size as well.

Android Firefox is the only browser to use native font scaling.


On May 30, 2021 at 2:04 AM, Manpreet Singh Sandhu ***@***.***) wrote:
"font:-apple-system-body" this solution not working for android chrome

**Comment by jefferyto** [comment]
> Looks like Apple is unwilling to change User Agent Stylesheet at the moment as well.

At the default system text size, `-apple-system-body` appears to set the font size to 17px. I imagine even a 1px difference (from the assumed 16px default) may break a lot of layouts in the wild. This may also make it untenable to include as a default for any major CSS framework.

**Comment by chrislachance** [comment]
Unfortunately, you may be right. But doing it anyways may be the right thing to do.

**Comment by tedw** [comment]
FWIW here’s is how I have been doing this for the past few years:

```scss
// Support iOS accessibility text size
// Note: Only apply on iOS due to bug with Safari 14 on Big Sur
// https://webkit.org/blog/3709/using-the-system-font-in-web-content/
// https://www.interactiveaccessibility.com/blog/text-resizing-web-pages-ios-using-dynamic-type
// https://dev.to/colingourlay/how-to-support-apple-s-dynamic-text-in-your-web-content-with-css-40c0
// https://gist.github.com/colingourlay/d95908ec5cd4854c7a5afa06f3989479
&.ua-ios {
  @supports (font: -apple-system-body) {
    font: -apple-system-body;
  }
}
```

FYI the `ua-ios` class is added to the `<html>` tag via a small bit of inline JS in the `<head>` (see example below) but you could remove that if you aren’t supporting Safari 14 on Big Sur.

```js
// Detect iOS
// http://stackoverflow.com/questions/9038625/detect-if-device-is-ios/9039885#9039885
if (/iPad|iPhone|iPod/.test(navigator.userAgent) && !window.MSStream) {
  document.documentElement.classList.add('ua-ios');
}
```

FWIW I also use this bit of CSS to ensure form fields don’t zoom on iOS, which happens if the font size is below 16px.

```
// Prevent zoom on iOS
// https://css-tricks.com/16px-or-larger-text-prevents-ios-form-zoom/
// http://www.456bereastreet.com/archive/201212/ios_webkit_browsers_and_auto-zooming_form_controls/
.ua-ios {
  input,
  textarea,
  select {
    // Use max() to allow for overriding the size in component styles
    @supports (font-size: #{"max(1em, 16px)"}) {
      font-size: #{"max(1em, 16px)"};
    }
  }
}
```

The odd `#{"max(1em, 16px)"}` syntax is to stop Sass from trying to use its own min and max functions (see https://github.com/sass/sass/issues/2849).

NB: You could also rewrite this using `:where()` to reduce specificity.
